pyat/developers.rst

```diff
@@ -114,24 +120,77 @@ To upload a release to PyPI, you will need to be a 'maintainer' of
 For testing any version that you have installed, the simple snippet in
 ``README.rst`` is sufficient.
 
-* Decide the Python versions that should be supported in the release
-   * Set these Python versions in ``python_requires`` in ``setup.cfg``
-   * Set at least these Python versions as ``python-version`` in ``.github/workflows/python-tests.yml``
-* Determine the minimum Numpy version that is required for those Python versions
-   * Set this numpy version in ``install_requires`` in ``setup.cfg``
-* Push a tag ``pyat-x.y.z`` to Github
+Decide the Python versions that should be supported in the release
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Set these Python versions in ``requires-python`` in ``pyproject.toml``
+* Set at least these Python versions as ``python-version`` in ``.github/workflows/python-tests.yml``
+
+Determine the minimum ``numpy`` and ``scipy`` versions:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* the version ensuring the requirements necessary to **run** PyAT is set in the
+  ``dependencies`` item of the ``[project]`` section of ``pyproject.toml``
+* The version required to **build** PyAT is set in the ``requires`` item of the
+  ``[build]`` section of ``pyproject.toml``. It depends on the python version and
+  must be higher or equal to the "run" version.
```

Comment:
I think this statement is confusing, and might be wrong. I haven't looked at this for a while.

As I understand it, the `dependencies` version needs to be the _lowest_ version specified in `requires`. This is because the numpy versions in `requires` are the lowest versions of `numpy` supported by each version of Python.

Reply:
I think the statement is right.
The version in `dependencies` is the one which includes all the functionalities needed by AT. It is by far anterior to the versions required to build, which depend on the python version. You are right that it would make no difference to increase it to the minimum required for building. But the logic is correct: the value in `dependencies` is based only on available functionalities in numpy.

Would you prefer increasing it to 1.16.6 ? No problem for me!

Reply:
It already is 1.16.6, I don't understand that last statement.

You still need ABI compatibility. This is in addition to the basic numpy functionalities required by AT. The build is performed with the same or equivalent versions as in requires, and so it should _never_ be higher than the "run" version.

Reply:
> It already is 1.16.6, I don't understand that last statement.

Sorry, thought it was just 1.16… 

> You still need ABI compatibility

ABI compatibility is a problem of compilation. It has to do with the build sequence, not to the functionality. So it makes no sense to have a `dependencies` specification higher that the `requires`: it would make no sense to be able the build with 1.16 if at run time you need a function added in 1.18 !

One the other hand, dependencies >= 1.16.6 means that any such version provides all the needed functionalities. If you are running python 3.10, it's well below the requirement for building, but that's how it is !

I confirm the statement: **the `dependencies` version must be lower or equal to the minimum build requirement in `requires`**. For instance, raising `dependencies` to 1.22 would break the compatibility with python 3.7 because 1.22 cannot be installed on it.

Reply:
> So it makes no sense to have a dependencies specification higher that the requires

This is correct. My disagreement is with whether you should have a `dependencies` specification _lower_ than the minimum in `requires`.

If your module is only ever built on versions 1.16 or above, you need to use numpy 1.16 (via `dependencies`) for ABI compatibility. As soon as you build on a later version, you break use with all lower versions of numpy.

This is fine for later versions of Python, because we always select the lowest version of numpy available for the given platform, in `requires`.

My view is that the dependencies version should always be _equal_ to the lowest `requires` version, and I don't think you've given a reason why we would want to avoid that.

Reply:
> My view is that the dependencies version should always be equal to the minimum requires version

That's how it is now, so no problem !

But if any time we decide to drop python 3.7, for some reason, I still don't see any reason to increase `dependencies` to 1.17: there is no change in the code, nor in the functionalities we use. Of course, when building, a version >=1.17 will be installed, but that's due to interactions between numpy and python, not our responsibility. As long as it's higher than 1.16.6, we are happy.

One could even say keeping 1.16.6 gives an indication of which of numpy functionalities used in AT…

Reply:
I don't disagree with your comments on how this works.

I think the documentation doesn't describe our full motivations properly. My preference is also to have 'stricter' rules about how we do things, if only to reduce our possible choices. My argument would also be invalid if we didn't restrict the Python version, i.e. saying it doesn't need to be >= Python 3.7.

One thing all of this relies on is the `requires` version being the absolute lowest supported by that version of Python, while still being above the version in `dependencies`. This is sort of explained in the following:

> To avoid version conflicts with the user's existing libraries, the pre-compiled binaries are built with the exact minimum library versions. 

Is this trying to refer to the ABI compatibility issue? Or something else? We could rephrase it to be more explicit:

> To avoid ABI compatibility issues, the pre-compiled binaries are built with the earliest possible version of numpy for the given Python version.

Reply:
Your last rephrasing looks perfect to me!